src/ImageSharp/Formats/Qoi/QoiEncoderCore.cs

```diff
@@ -0,0 +1,222 @@
+                     * black. However, the decoder of this project takes that into consideration
+                     *
+                     * To further details, see https://github.com/phoboslab/qoi/issues/258,
+                     * and we should discuss what to do about this approach and
+                     * if it's correct
+                     */
+                    byte repetitions = 0;
+                    do
+                    {
+                        repetitions++;
+                        j++;
+                        if (j == row.Length)
+                        {
+                            j = 0;
+                            i++;
+                            if (i == pixels.Height)
+                            {
+                                break;
+                            }
+
+                            row = pixels.DangerousGetRowSpan(i);
+                        }
+
+                        currentPixel = row[j];
+                        currentPixel.ToRgba32(ref currentRgba32);
+                    }
+                    while (currentRgba32.Equals(previousPixel) && repetitions < 62);
+
+                    j--;
+                    stream.WriteByte((byte)((byte)QoiChunk.QoiOpRun | (repetitions - 1)));
+
+                    /* If it's a QOI_OP_RUN, we don't overwrite the previous pixel since
+                     * it will be taken and compared on the next iteration
+                     */
+                    continue;
+                }
+
+                // else, we check if it exists in the previously seen pixels
+                // If so, we do a QOI_OP_INDEX
+                int pixelArrayPosition = GetArrayPosition(currentRgba32);
+                if (previouslySeenPixels[pixelArrayPosition].Equals(currentPixel))
+                {
+                    stream.WriteByte((byte)pixelArrayPosition);
+                }
+                else
+                {
+                    // else, we check if the difference is less than -2..1
+                    // Since it wasn't found on the previously seen pixels, we save it
+                    previouslySeenPixels[pixelArrayPosition] = currentRgba32;
+
+                    sbyte diffRed = (sbyte)(currentRgba32.R - previousPixel.R);
+                    sbyte diffGreen = (sbyte)(currentRgba32.G - previousPixel.G);
+                    sbyte diffBlue = (sbyte)(currentRgba32.B - previousPixel.B);
+
+                    // If so, we do a QOI_OP_DIFF
+                    if (diffRed is >= -2 and <= 1 &&
+                        diffGreen is >= -2 and <= 1 &&
+                        diffBlue is >= -2 and <= 1 &&
+                        currentRgba32.A == previousPixel.A)
+                    {
+                        // Bottom limit is -2, so we add 2 to make it equal to 0
+                        byte dr = (byte)(diffRed + 2);
+                        byte dg = (byte)(diffGreen + 2);
+                        byte db = (byte)(diffBlue + 2);
```

Comment:
Unless you need to clamp the numbers, can't you `int` instead of `sbyte` and `byte` until the point when you actually need to write a `byte` to the stream or a assign it to a value within a pixel? That would get rid of the conversions, making things faster.

This suggestion applies to all code in both the Decoder & Encoder.

Reply:
Oh okay, I though I had to save memory as much as I could, but yeah, I also prefer use ints to avoid conversions xD

Reply:
> I had to save memory

The JIT will map these variables to CPU registers. Memory is a concern when you work with larger buffers, not small local variables.